django/db/models/sql/compiler.py

```diff
@@ -223,7 +223,12 @@ def get_select_from_parent(klass_info):
 
         ret = []
         for col, alias in select:
-            ret.append((col, self.compile(col, select_format=True), alias))
+            try:
+                sql, params = self.compile(col, select_format=True)
+            except EmptyResultSet:
+                # Selecting a predicate that is always false:
+                sql, params = '0', ()
```

Comment:
The same try/except fixes `When` (see last commit). I'll try to push a cleaned up commit tonight (UTC+1).